Engine/Helper.cs

```diff
@@ -1323,156 +1323,177 @@ internal List<RuleSuppression> GetSuppressionsConfiguration(ConfigurationDefinit
-                unSuppressedRecords.AddRange(diagnostics);
+                unsuppressedRecords.AddRange(diagnostics);
                 return result;
             }
 
-            List<RuleSuppression> ruleSuppressions = ruleSuppressionsDict[ruleName];
-            var offsetArr = GetOffsetArray(diagnostics);
-            int recordIndex = 0;
-            int startRecord = 0;
-            bool[] suppressed = new bool[diagnostics.Count];
-            foreach (RuleSuppression ruleSuppression in ruleSuppressions)
+            // We need to report errors for any rule suppressions with IDs that are not applied to anything
+            // Start by assuming all suppressions are unapplied and then we'll remove them as we apply them later
+            var unappliedSuppressions = new HashSet<RuleSuppression>();
+            foreach (RuleSuppression suppression in ruleSuppressions)
             {
-                int suppressionCount = 0;
-                while (startRecord < diagnostics.Count
-                    // && diagnostics[startRecord].Extent.StartOffset < ruleSuppression.StartOffset)
-                    // && diagnostics[startRecord].Extent.StartLineNumber < ruleSuppression.st)
-                    && offsetArr[startRecord] != null && offsetArr[startRecord].Item1 < ruleSuppression.StartOffset)
+                if (!string.IsNullOrWhiteSpace(suppression.RuleSuppressionID))
                 {
-                    startRecord += 1;
+                    unappliedSuppressions.Add(suppression);
                 }
-
-                // at this point, start offset of startRecord is greater or equals to rulesuppression.startoffset
-                recordIndex = startRecord;
-
-                while (recordIndex < diagnostics.Count)
+            }
+            
+            // We have suppressions and diagnostics
+            // For each diagnostic, collect all the suppressions that apply to it
+            // and if there are any, record the diagnostic as suppressed
+            foreach (DiagnosticRecord diagnostic in diagnostics)
+            {
+                Tuple<int, int> offsetPair = GetOffset(diagnostic);
+                
+                var appliedSuppressions = new List<RuleSuppression>();
+                foreach (RuleSuppression suppression in ruleSuppressions)
                 {
-                    DiagnosticRecord record = diagnostics[recordIndex];
-                    var curOffset = offsetArr[recordIndex];
-
-                    //if (record.Extent.EndOffset > ruleSuppression.EndOffset)
-                    if (curOffset != null && curOffset.Item2 > ruleSuppression.EndOffset)
+                    // Check that the diagnostic is within the suppressed extent, if we have such an extent
+                    if (!(offsetPair is null)
+                        && (offsetPair.Item1 < suppression.StartOffset || offsetPair.Item2 > suppression.EndOffset))
                     {
-                        break;
+                        continue;
```

Comment:
Is this intentional? Just to double check.

Reply:
Yeah, in the old logic we'd take each suppression and look for all the diagnostics it applied to (so if diagnostics are emitted in extent order, we could stop once we'd seen the first diagnostic after the suppression's extent).

In the new logic we take each diagnostic and look for all the suppressions that apply to it, so we can't stop early.